portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,34 +14,54 @@
 
 package com.google.sps.servlets;
 
-import java.util.Map;
-import java.util.HashMap;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import com.google.gson.Gson;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
+// TODO: Only allow unique usernames - returning users should still be able to write additional comments with the same username
 /** Servlet that returns inputted fun facts from users and stores previously inputted facts. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  private static final String FIRST_NAME_KEY = "first-name";
-  private static final String LAST_NAME_KEY = "last-name";
+  private static final String USER_NAME_KEY = "user-name";
   private static final String FUN_FACT_KEY = "user-fun-fact";
   private static final String NO_NAME = "Anonymous";
-  private Map<String, String> oldFacts = new HashMap<String, String>();
+  private static final String ENTITY_NAME = "FunFact";
+  private static final String PROPERTY_NAME = "name";
+  private static final String PROPERTY_FACT = "fact";
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String oldFactsJson = convertToJsonUsingGson(oldFacts);
+    Query query = new Query(ENTITY_NAME).addSort(PROPERTY_FACT, SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    // Add previously inputted fun facts to userFacts
+    ImmutableList<String> userFacts =
+        Streams.stream(results.asIterable())
+        .map(entity -> entity.getProperty(PROPERTY_NAME).toString() + ": " + entity.getProperty(PROPERTY_FACT).toString())
```

Comment:
nit: let's put this lamda method (the logic after the arrow) into a helper method that takes an entity and returns a String. Then we can simply reference that helper method in the map operator by calling .map(DataServlet::methodName), making it a lot easier to read. "methodName" refers to the name of the helper method we're defining, and "DataServlet" is the class it's defined in (in this case, since you'll define the helper method in this file/class, then we use the name of the class "DataServlet")

note: on the .map(DataServlet::methodName) syntax. This is equivalent to .map(entity->methodName(entity)) but is the preferred style. This syntax only works on single-argument methods where the argument type is equivalent to the type of the "object" reference in our stream (eg: "entity" has type Entity and our helper method will only take a single Entity, so we can use this syntax. If our helper method took multiple arguments (say an Entity _and_ a String), then we can't use this syntax)

Reply:
Gotcha, thanks for bringing this up! This definitely makes the code easier to read